### ISSUE TITLE
Add TypeScript support

### ISSUE THREAD

**Issue by nullndr** [juanjoDiaz/json2csv#1]
Hey I just saw the deprecation notice [here](https://github.com/zemirco/json2csv/pull/575).

I wanted to know if you were planning to add TypeScript directly here instead of use [Definitely Typed](https://github.com/DefinitelyTyped/DefinitelyTyped).

**Comment by juanjoDiaz** [comment]
Hi @nullndr ,

It is in my TODO list.
I wished to have it for v6. Unfortunately, CSV and JSON are pretty open formats and it will take quite some effort to move things to typescript.

However, now that this library is broken into smaller modules it might be much easier to do the migration.
In the meantime, I'll try to publish the types to [Definitely Typed](https://github.com/DefinitelyTyped/DefinitelyTyped).

**Comment by nullndr** [comment]
Good, thank you

**Comment by soullivaneuh** [comment]
@juanjoDiaz @nullndr I think this issue should be kept opened until it is completely resolved for a better tracking, is that possible to change the status?

Typescript support is becoming a kind of "Must have" for a JavaScript library now, I totally support this request. This is currently the only blocking part to be able to make the upgrade to v6. :+1:

**Comment by nullndr** [comment]
@soullivaneuh yeah not a problem for me

**Comment by juanjoDiaz** [comment]
I agree.
This is crucial and I'm working hard to migrate the library to TS.
This is not trivial because JSON and CSV are both dynamic formats and json2csv needs to stay totally generic.
But it's coming.

**Comment by nullndr** [comment]
@juanjoDiaz if you like I could help you with typing the packages

**Comment by helciofranco** [comment]
I could help as well.

Currently, I'm using just a single part of `@json2csv/plainjs` package, and I did a very simple type definition:

```tsx
declare module '@json2csv/plainjs' {
  export class Parser {
    parse: <T>(data: T) => BlobPart
  }
}
```

**Comment by say8425** [comment]
```ts
declare module "@json2csv/plainjs" {
  export type FieldValueCallback =
    | FieldValueCallbackWithoutField
    | FieldValueCallbackWithField;

  export interface FieldValueCallbackWithoutField {
    (row: any): any;
  }

  export interface FieldValueCallbackWithField {
    (row: T, field: FieldValueCallbackInfo): any;
  }
  export interface FieldInfo {
    label?: string;
    default?: any;
    value: any | FieldValueCallback;
  }

  export type Flatten = {
    objects: true;
    arrays: false;
    separator: ".";
  };

  export type Unwind = {
    paths: string[];
    blankout: false;
  };

  export type Transform = {
    flatten?: Flatten;
    unwind?: Unwind;
  };

  export type Formatters = {
    undefined?: any;
    boolean?: any;
    number?: any;
    bigint?: any;
    string?: any;
    symbol?: any;
    function?: any;
    object?: any;
    headers?: any;
  };

  export type Options = {
    fields?: Array<string | FieldInfo<T>>;
    transforms?: Transform[];
    formatters?: Formatters;
    defaultValue?: any;
    delimiter?: ",";
    eol?: string;
    header?: true;
    includeEmptyRows?: false;
    withBOM?: false;
  };
  export class Parser {
    constructor(options?: Options);

    parse: <T>(data: T[]) => BlobPart;
  }
}
```

I code and use this, not perfect.

**Comment by micalevisk** [comment]
the ambient file I've been using so far for `@json2csv/node` and `@json2csv/transforms`

<details>
<summary><code>global.d.ts</code></summary>

```ts
declare module '@json2csv/node' {
  import { Transform as CoreTransform, TransformOptions } from 'stream';

  interface FieldValueCallbackInfo {
    label: string;
    default?: string | undefined;
  }

  type FieldValueCallback<T> =
    | FieldValueCallbackWithoutField<T>
    | FieldValueCallbackWithField<T>;

  interface FieldInfo<T> {
    label?: string | undefined;
    default?: string | undefined;
    value: string | FieldValueCallback<T>;
  }

  interface NormalizedFieldInfo<T> {
    label: string;
    value: FieldValueCallback<T>;
  }

  interface Options<T> {
    fields?: Array<string | FieldInfo<T>> | undefined;
    ndjson?: boolean | undefined;
    defaultValue?: string | undefined;
    quote?: string | undefined;
    escapedQuote?: string | undefined;
    delimiter?: string | undefined;
    eol?: string | undefined;
    excelStrings?: boolean | undefined;
    header?: boolean | undefined;
    includeEmptyRows?: boolean | undefined;
    withBOM?: boolean | undefined;
    transforms?: Array<Json2CsvTransform<any, any>> | undefined;
  }

  export class Transform<T> extends CoreTransform {
    constructor(opts?: Options<T>, transformOpts?: TransformOptions);
  }
}

declare module '@json2csv/transforms' {
  interface UnwindOptions {
    paths: string[] | undefined;
    /** @default false */
    blankOut?: boolean | undefined;
  }

  interface FlattenOptions {
    /** @default true */
    objects?: boolean | undefined;
    /** @default false */
    arrays?: boolean | undefined;
    /** @default '.'' */
    separator?: string | undefined;
  }

  /**
   * Builds a unwinding transform
   *
   * @param options Options to use for unwinding
   * @returns Array of objects containing all rows after unwinding of chosen paths
   */
  export declare function unwind(
    options?: UnwindOptions,
  ): Json2CsvTransform<any, any[]>;

  /**
   * Builds a flattening transform
   *
   * @param options Options to use for flattening
   * @returns Flattening transform
   */
  export declare function flatten(
    options?: FlattenOptions,
  ): Json2CsvTransform<anu, any>;
}
```

</details>

**Comment by juanjoDiaz** [comment]
Hi guys,

Sorry for this taking so long...
I already have the library migrated to Typescript. You can find it on https://github.com/juanjoDiaz/json2csv/tree/migrate_to_ts
I'm now trying to migrate the tests to Typescript. You can see the works on https://github.com/juanjoDiaz/json2csv/tree/migrate_to_ts_tests

I think that, for now, you could import on npm directly from the migrate_to_ts branch.

Also, any help to migrate the tests to TS and complete the migration is more than welcome.

**Comment by nullndr** [comment]
@juanjoDiaz I can help you with the migration of tests if you would like

**Comment by shayc** [comment]
In case anyone wants to install and import the TS version of json2csv:

With Yarn:
`yarn add juanjoDiaz/json2csv#migrate_to_ts`

With NPM:
`npm install juanjoDiaz/json2csv#migrate_to_ts`

**Comment by juanjoDiaz** [comment]
Thanks @shayc and please guys let me know if you find any issue 🙂

**Comment by sharon182** [comment]
@shayc @juanjoDiaz 
I'm trying to use the `migrate_to_ts` branch but it's still complaining about missing declaration file.
`yarn add json2csv#migrate_to_ts`
What am I missing ? 
Thanks !

**Comment by juanjoDiaz** [comment]
I just released v7.0.0 which entirely rewrites @json2csv to typescript.

Let me know if you find any issues with it. This was a massive change.
